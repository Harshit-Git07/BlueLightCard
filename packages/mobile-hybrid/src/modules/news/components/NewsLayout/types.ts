import { NewsModel } from '@/models/news';

export interface NewsLayoutProps {
  news: NewsModel[];
  showHeading?: boolean;
  onClickSeeAll?: () => void;
}
