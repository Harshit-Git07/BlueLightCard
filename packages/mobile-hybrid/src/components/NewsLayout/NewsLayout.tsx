import dayjs from 'dayjs';
import { NewsLayoutProps } from './types';
import Heading from '@/components/Heading/Heading';
import { cssUtil } from '@/utils/cssUtil';
import ListItem from '@/components/ListItem/ListItem';
import { FC } from 'react';

export const formatDate = (date: string) => {
  return dayjs(date, 'YYYY-MM-DD HH:mm:ss').format('ddd DD MMM, YYYY');
};

const NewsLayout: FC<NewsLayoutProps> = ({
  news,
  showHeading = true,
  onClickSeeAll,
  onArticleClick,
}) => {
  return (
    <>
      {showHeading && <Heading title="Latest news" onClickSeeAll={onClickSeeAll} />}
      {!!news.length && (
        <ul>
          {news.map((article, index) => {
            const classes = cssUtil([
              'py-3',
              index < news.length - 1
                ? 'border-b border-listItem-divider-colour-light dark:border-listItem-divider-colour-dark'
                : '',
            ]);
            return (
              <li className={classes} key={article.newsId}>
                <ListItem
                  className="ml-4"
                  title={article.title}
                  text={formatDate(article.when)}
                  imageSrc={article.s3image}
                  onClick={() => onArticleClick(article.newsId)}
                />
              </li>
            );
          })}
        </ul>
      )}
    </>
  );
};

export default NewsLayout;
