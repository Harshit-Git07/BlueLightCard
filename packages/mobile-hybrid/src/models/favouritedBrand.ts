export interface FavouritedBrandsModel {
  cid: number;
  companyname: string;
  logos: string;
}
