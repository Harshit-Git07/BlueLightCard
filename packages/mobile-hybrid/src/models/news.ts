export interface NewsModel {
  nid: number;
  newsId: string;
  when: string;
  title: string;
  body: string;
  image: string;
  s3image: string;
}
