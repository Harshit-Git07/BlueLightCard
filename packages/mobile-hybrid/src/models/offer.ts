export interface OfferPromosModel {
  id: number;
  compid: number;
  offername: string;
  logos: string;
  companyname: string;
  image: string;
  s3Image: string;
  absoluteLogos: string;
  s3logos: string;
  absoluteImage: string;
}

export interface OfferSharedModel {
  title: string;
  random: boolean;
  items: OfferPromosModel[];
}

export interface OfferFlexibleModel {
  title: string;
  subtitle: string;
  random: boolean;
  items: {
    id: number;
    title: string;
    imagehome: string;
    imagedetail: string;
    navtitle: string;
    intro: string;
    footer: string;
    random: boolean;
    hide: boolean;
    items: any[];
  }[];
}

export interface OfferDataModel {
  deals: OfferSharedModel[];
  flexible: OfferFlexibleModel;
  groups: OfferSharedModel[];
}
