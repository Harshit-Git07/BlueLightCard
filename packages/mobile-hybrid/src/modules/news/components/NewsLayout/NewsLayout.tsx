import { FC } from 'react';
import dayjs from 'dayjs';
import { NewsLayoutProps } from './types';
import Heading from '@/components/Heading/Heading';
import { cssUtil } from '@/utils/cssUtil';
import ListItem from '@/components/ListItem/ListItem';
import InvokeNativeNavigation from '@/invoke/navigation';

const formatDate = (date: string) => {
  return dayjs(date, 'YYYY-MM-DD HH:mm:ss').format('ddd DD MMM, YYYY');
};

const navigation = new InvokeNativeNavigation();

const NewsLayout: FC<NewsLayoutProps> = ({ news, showHeading = true, onClickSeeAll }) => {
  return (
    <>
      {showHeading && <Heading title="Latest news" onClickSeeAll={onClickSeeAll} />}
      {!!news.length && (
        <ul>
          {news.map((article, index) => {
            const classes = cssUtil([
              'py-3',
              index < news.length - 1 ? 'border-b dark:border-neutral-grey-100' : '',
            ]);
            return (
              <li className={classes} key={article.newsId}>
                <ListItem
                  className="ml-4"
                  title={article.title}
                  text={formatDate(article.when)}
                  imageSrc={article.s3image}
                  onClick={() => navigation.navigate(`/news/${article.nid}`)}
                />
              </li>
            );
          })}
        </ul>
      )}
    </>
  );
};

export default NewsLayout;
