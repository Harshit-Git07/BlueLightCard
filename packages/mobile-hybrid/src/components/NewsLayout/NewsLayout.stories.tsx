import { Meta, StoryFn } from '@storybook/react';
import tokenMigrationDecorator from '@storybook/tokenMigrationDecorator';
import NewsLayout, { formatDate } from './NewsLayout';
import { NewsModel } from '@/models/news';

const componentMeta: Meta = {
  title: 'Layouts/NewsLayout',
  component: NewsLayout,
  decorators: [tokenMigrationDecorator],
  argTypes: {
    newsOne: {
      name: 'Show News Item One',
      control: 'boolean',
    },
    newsOneId: {
      name: 'News Item One ID',
      control: 'string',
      if: {
        arg: 'newsOne',
      },
    },
    newsOneDate: {
      name: 'News Item One Date',
      control: 'date',
      if: {
        arg: 'newsOne',
      },
    },
    newsOneTitle: {
      name: 'News Item One Title',
      control: 'string',
      if: {
        arg: 'newsOne',
      },
    },
    newsOneBody: {
      name: 'News Item One Body',
      control: 'string',
      if: {
        arg: 'newsOne',
      },
    },
    newsOneImage: {
      name: 'News Item One Image',
      control: 'string',
      if: {
        arg: 'newsOne',
      },
    },
    newsTwo: {
      name: 'Show News Item Two',
      control: 'boolean',
    },
    newsTwoId: {
      name: 'News Item Two ID',
      control: 'string',
      if: {
        arg: 'newsTwo',
      },
    },
    newsTwoDate: {
      name: 'News Item Two Date',
      control: 'date',
      if: {
        arg: 'newsTwo',
      },
    },
    newsTwoTitle: {
      name: 'News Item Two Title',
      control: 'string',
      if: {
        arg: 'newsTwo',
      },
    },
    newsTwoBody: {
      name: 'News Item Two Body',
      control: 'string',
      if: {
        arg: 'newsTwo',
      },
    },
    newsTwoImage: {
      name: 'News Item Two Image',
      control: 'string',
      if: {
        arg: 'newsTwo',
      },
    },
  },
  parameters: {
    controls: {
      exclude: ['news', 'onClickSeeAll', 'onArticleClick'],
    },
  },
};

const DefaultTemplate: StoryFn = ({
  showHeading,
  newsOne,
  newsOneId,
  newsOneDate,
  newsOneTitle,
  newsOneBody,
  newsOneImage,
  newsTwo,
  newsTwoId,
  newsTwoDate,
  newsTwoTitle,
  newsTwoBody,
  newsTwoImage,
}) => {
  const news: NewsModel[] = [];
  if (newsOne) {
    news.push({
      nid: 1,
      body: newsOneBody,
      image: newsOneImage,
      s3image: newsOneImage,
      newsId: newsOneId,
      title: newsOneTitle,
      when: formatDate(newsOneDate),
    });
  }
  if (newsTwo) {
    news.push({
      nid: 2,
      body: newsTwoBody,
      image: newsTwoImage,
      s3image: newsTwoImage,
      newsId: newsTwoId,
      title: newsTwoTitle,
      when: formatDate(newsTwoDate),
    });
  }
  return (
    <NewsLayout
      news={news}
      onArticleClick={(articleid) => alert(`Article with ID: ${articleid} clicked!`)}
      onClickSeeAll={() => alert(`See All Clicked!`)}
      showHeading={showHeading}
    />
  );
};

export const Default = DefaultTemplate.bind({});

Default.args = {
  showHeading: true,
  newsOne: true,
  newsOneId: 'news-one-id',
  newsOneDate: '2023-12-12',
  newsOneTitle: 'News One Title',
  newsOneBody: 'This is the news one body text',
  newsOneImage: 'https://picsum.photos/600/250',
  newsTwo: false,
  newsTwoId: '',
  newsTwoDate: '',
  newsTwoTitle: '',
  newsTwoBody: '',
  newsTwoImage: '',
};

export default componentMeta;
