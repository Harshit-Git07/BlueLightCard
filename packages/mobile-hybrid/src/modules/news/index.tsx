import { FC } from 'react';
import useNews from '@/hooks/useNews';
import NewsLayout from './components/NewsLayout/NewsLayout';
import { newsPanelStore } from './store';
import InvokeNativeNavigation from '@/invoke/navigation';
import { useSetAtom } from 'jotai';

const navigation = new InvokeNativeNavigation();

/**
 * Preview news articles
 */
const NewsPreview: FC = () => {
  const setSeeAllNews = useSetAtom(newsPanelStore);

  const news = useNews(true);

  const onSeeAllClick = () => {
    document.getElementById('app-body')?.classList.add('noscroll');
    setSeeAllNews(true);
  };

  return (
    <NewsLayout
      news={news}
      onClickSeeAll={onSeeAllClick}
      onArticleClick={(articleId) =>
        navigation.navigate(`/bluelightcardnewsdetails.php?id=${articleId}`, 'home')
      }
    />
  );
};

/**
 * Lists all news articles
 */
const NewsList: FC = () => {
  const news = useNews();

  return (
    <NewsLayout
      news={news}
      showHeading={false}
      onArticleClick={(articleId) =>
        navigation.navigate(`/bluelightcardnewsdetails.php?id=${articleId}`, 'home')
      }
    />
  );
};

export { NewsPreview, NewsList };
