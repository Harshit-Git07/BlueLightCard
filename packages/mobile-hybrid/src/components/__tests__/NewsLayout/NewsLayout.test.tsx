import NewsLayout, { formatDate } from '@/components/NewsLayout/NewsLayout';
import { NewsModel } from '@/models/news';
import { fireEvent, render, screen } from '@testing-library/react';

// Mocking Next Image allows us to input a fake url for testing purposes
jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line @next/next/no-img-element, jsx-a11y/alt-text
    return <img {...props} alt="" />;
  },
}));

const mockNews: NewsModel[] = [
  {
    body: 'This is the news body',
    image: 'image/url',
    s3image: 's3/image/url',
    newsId: 'newsId1234',
    nid: 1,
    title: 'News Title',
    when: '2022-12-9',
  },
];

describe('NewsLayout.tsx', () => {
  describe('formatDate', () => {
    it('should return a string date formatted correctly', () => {
      const date = '2022-12-9';
      expect(formatDate(date)).toStrictEqual('Fri 09 Dec, 2022');
    });
    it('should return invalid date if an empty string is passed', () => {
      const emptyString = '';
      expect(formatDate(emptyString)).toStrictEqual('Invalid Date');
    });
  });

  it('renders, showing the heading by default', () => {
    const { container } = render(
      <NewsLayout news={mockNews} onArticleClick={jest.fn()} onClickSeeAll={jest.fn()} />,
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Latest news')).toBeDefined();
  });

  it('should not show the heading when set to false', () => {
    render(
      <NewsLayout
        news={mockNews}
        onArticleClick={jest.fn()}
        onClickSeeAll={jest.fn()}
        showHeading={false}
      />,
    );
    expect(screen.queryByText('Latest news')).toBeNull();
  });

  it('should call the onClickSeeAll function when the heading is clicked', async () => {
    const onClick = jest.fn();
    render(<NewsLayout news={mockNews} onArticleClick={jest.fn()} onClickSeeAll={onClick} />);
    const heading = screen.getByText('See all');
    await fireEvent.click(heading);
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should call the onArticleClick with the id of the article', async () => {
    const onArticleClick = jest.fn();
    render(
      <NewsLayout news={mockNews} onArticleClick={onArticleClick} onClickSeeAll={jest.fn()} />,
    );
    const newsArticle = screen.getByText(mockNews[0].title);
    await fireEvent.click(newsArticle);
    expect(onArticleClick).toHaveBeenCalledWith(mockNews[0].newsId);
  });
});
