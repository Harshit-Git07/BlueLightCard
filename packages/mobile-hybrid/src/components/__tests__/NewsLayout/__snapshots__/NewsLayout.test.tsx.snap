// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsLayout.tsx renders, showing the heading by default 1`] = `
<div>
  <div
    class="flex w-full content-center my-2 px-4"
  >
    <h2
      class="font-heading-font font-heading-font-weight text-heading-font tracking-heading-font leading-heading-font flex-1 text-heading-colour-light dark:text-heading-colour-dark"
    >
      Latest news
    </h2>
    <button
      class="font-heading-link-font font-heading-link-font-weight text-heading-link-font tracking-heading-link-font leading-heading-link-font text-heading-link-colour dark:text-heading-link-colour-dark"
    >
      See all
    </button>
  </div>
  <ul>
    <li
      class="py-3 "
    >
      <div
        class="flex ml-4 cursor-pointer"
        role="button"
      >
        <div
          class="relative md:w-[28%] md:pb-[14%] w-[50%] pb-[20%]"
        >
          <img
            alt=""
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="sm:object-cover object-contain rounded"
            placeholder="blur"
            src="s3/image/url"
          />
        </div>
        <div
          class="flex flex-col justify-center pl-3 w-full"
        >
          <h4
            class="text-listItem-title-colour-light dark:text-listItem-title-colour-dark text-listItem-title-font font-listItem-title-font font-listItem-title-font-weight mb-1 tracking-listItem-title-font leading-listItem-title-font line-clamp-1 md:line-clamp-2"
          >
            News Title
          </h4>
          <p
            class="text-listItem-text-colour-light dark:text-listItem-text-colour-dark text-listItem-text-font font-listItem-text-font font-listItem-text-font-weight tracking-listItem-text-font leading-listItem-text-font lg:text-lg"
          >
            Fri 09 Dec, 2022
          </p>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
